external-crates/move/crates/move-compiler/src/sui_mode/linters/public_random.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This analysis flags uses of random::Random and random::RandomGenerator in public functions.
+
+use crate::diagnostics::WarningFilters;
+use crate::expansion::ast::{Address, ModuleIdent};
+use crate::parser::ast::FunctionName;
+use crate::typing::visitor::{TypingVisitorConstructor, TypingVisitorContext};
+use crate::{
+    diag,
+    diagnostics::codes::{custom, DiagnosticInfo, Severity},
+    expansion::ast::{ModuleIdent_, Visibility},
+    naming::ast as N,
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    typing::ast as T,
+};
+
+use super::{
+    LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX,
+    RANDOM_GENERATOR_STRUCT_NAME, RANDOM_MOD_NAME, RANDOM_STRUCT_NAME, SUI_PKG_NAME,
+};
+
+const PUBLIC_RANDOM_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::PublicRandom as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "Risky use of 'sui::random'",
+);
+
+pub struct PublicRandomVisitor;
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for PublicRandomVisitor {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn add_warning_filter_scope(&mut self, filter: WarningFilters) {
+        self.env.add_warning_filter_scope(filter)
+    }
+
+    fn pop_warning_filter_scope(&mut self) {
+        self.env.pop_warning_filter_scope()
+    }
+
+    fn visit_function_custom(
+        &mut self,
+        module: ModuleIdent,
+        fname: FunctionName,
+        fdef: &mut T::Function,
+    ) -> bool {
+        if fdef.attributes.is_test_or_test_only()
+            || !matches!(fdef.visibility, Visibility::Public(_))
+            || is_sui_module(module)
+        {
+            return false;
```

Comment:
We don't need to visit the function body, so these can all be true
```suggestion
            return true;
```